src/transaction/locator_sr.c

```diff
@@ -13937,6 +13937,76 @@ has_errors_filtered_for_insert (std::vector<int> error_filter_array)
 }
 // *INDENT-ON*
 
+void
+locator_change_classname_entry_status_if_exists (THREAD_ENTRY * thread_p, const OID * classoid,
+						 const LC_FIND_CLASSNAME status)
+{
+  assert (is_passive_transaction_server ());
+
+  char *classname = NULL;
+
+  (void) heap_get_class_name (thread_p, classoid, &classname);
+  if (classname != NULL)
+    {
+      if (csect_enter (thread_p, CSECT_LOCATOR_SR_CLASSNAME_TABLE, INF_WAIT) != NO_ERROR)
+	{
+	  assert (false);
+	  return;
+	}
+
+      LOCATOR_CLASSNAME_ENTRY *entry = (LOCATOR_CLASSNAME_ENTRY *) mht_get (locator_Mht_classnames, classname);
+      if (entry != NULL)
+	{
+	  assert (entry->e_current.action != status);
+
+	  entry->e_current.action = status;
+	}
+      else
+	{
+	  /* CREATEd or RENAMEd class is not cached in the locator_Mht_classnames yet.
+	   * They are updated when replicator apply LOG_COMMIT.
+	   * scenario (PTS side):
+	   * 1) CREATE
+	   * CREATE TABLE tbl (a int);  -- "tbl" is not cached in the locator_Mht_classnames because it is not committed yet.
+	   * DROP TABLE tbl;            -- looking for "tbl" in the locator_Mht_classnames above, but not found.
+	   *                               We don't need to change the status of "tbl" entry,
+	   *                               it will be changed at LOG_COMMIT time
+	   * COMMIT;
+	   *
+	   * 2) RENAME
+	   * RENAME TABLE tbl to tbl2;  -- "tbl2" is not cached, but only "tbl" is cached
+	   * DROP TABLE tbl2;           -- looking for "tbl2" in the cache, but not found.
+	   *                               But we need to change the status of "tbl" entry
+	   *                               because READ transaction in PTS could access DELETED "tbl" object
+	   *                               using cached entry which causes failure.
+	   * COMMIT;
+	   */
+
+	  HENTRY_PTR hentry = NULL;
+	  HENTRY_PTR next = NULL;
+	  for (hentry = locator_Mht_classnames->act_head; hentry != NULL; hentry = next)
```

Comment:
This may seem like a rather expensive operation. If you agree, please leave a TODO comment for improvements.